pkg/kclient/generator/utils.go

```diff
@@ -0,0 +1,160 @@
+}
+
+// getResourceReqs creates a kubernetes ResourceRequirements object based on resource requirements set in the devfile
+func getResourceReqs(comp common.DevfileComponent) corev1.ResourceRequirements {
+	reqs := corev1.ResourceRequirements{}
+	limits := make(corev1.ResourceList)
+	if comp.Container != nil && comp.Container.MemoryLimit != "" {
+		memoryLimit, err := resource.ParseQuantity(comp.Container.MemoryLimit)
+		if err == nil {
+			limits[corev1.ResourceMemory] = memoryLimit
+		}
+		reqs.Limits = limits
+	}
+	return reqs
+}
+
+// addSyncRootFolder adds the sync root folder to the container env and volume mounts
+func addSyncRootFolder(container *corev1.Container, sourceMapping string) string {
+	var syncRootFolder string
+	if sourceMapping != "" {
+		syncRootFolder = sourceMapping
+	} else {
+		syncRootFolder = DevfileSourceVolumeMount
+	}
+
+	// Note: PROJECTS_ROOT & PROJECT_SOURCE are validated at the devfile parser level
+	// Add PROJECTS_ROOT to the container
+	container.Env = append(container.Env,
+		corev1.EnvVar{
+			Name:  EnvProjectsRoot,
+			Value: syncRootFolder,
+		})
+
+	return syncRootFolder
+}
+
+// addSyncFolder adds the sync folder path for the container
+// sourceVolumePath: mount path of the empty dir volume to sync source code
+// projects: list of projects from devfile
+func addSyncFolder(container *corev1.Container, sourceVolumePath string, projects []common.DevfileProject) error {
+	var syncFolder string
+
+	// if there are no projects in the devfile, source would be synced to $PROJECTS_ROOT
+	if len(projects) == 0 {
+		syncFolder = sourceVolumePath
+	} else {
+		// if there is one or more projects in the devfile, get the first project and check its clonepath
+		project := projects[0]
+		// If clonepath does not exist source would be synced to $PROJECTS_ROOT/projectName
+		syncFolder = filepath.ToSlash(filepath.Join(sourceVolumePath, project.Name))
+
+		if project.ClonePath != "" {
+			if strings.HasPrefix(project.ClonePath, "/") {
+				return fmt.Errorf("the clonePath %s in the devfile project %s must be a relative path", project.ClonePath, project.Name)
+			}
+			if strings.Contains(project.ClonePath, "..") {
+				return fmt.Errorf("the clonePath %s in the devfile project %s cannot escape the value defined by $PROJECTS_ROOT. Please avoid using \"..\" in clonePath", project.ClonePath, project.Name)
+			}
+			// If clonepath exist source would be synced to $PROJECTS_ROOT/clonePath
+			syncFolder = filepath.ToSlash(filepath.Join(sourceVolumePath, project.ClonePath))
+		}
+	}
+
+	container.Env = append(container.Env,
+		corev1.EnvVar{
+			Name:  EnvProjectsSrc,
+			Value: syncFolder,
+		})
+
+	return nil
+}
+
+// GetPortExposure iterate through all endpoints and returns the highest exposure level of all TargetPort.
+// exposure level: public > internal > none
+// This function should be under parser pkg
+func GetPortExposure(containerComponents []common.DevfileComponent) map[int32]common.ExposureType {
```

Comment:
shouldn't this function and the ones below be non util functions?

Reply:
So 

1. funcs which have devfile obj as input and output should be going to the parser pkg in devfile/library; dont want to move it there in odo now since that is already merged to devfile/library

2. funcs that return kube obj should remain in utils and esp if they're only used by the generator funcs

Reply:
But `GetDevfileContainerComponents()` and `GetDevfileVolumeComponents()` are returning devfile objects and are being used by odo. Even `GetPortExposure()` is being used by odo.

Reply:
So 

- `GetDevfileContainerComponents()`, `GetDevfileVolumeComponents()` & `GetPortExposure()` will be in the parser pkg in devfile/library and not in the current generator pkg because they return devfile objects and kube structs. But I dont want to move it to the parser pkg in odo because that code is already copied to devfile/library.. I've put a comment in the function desc saying it will be moved to parser pkg.

-  `convertEnvs()`, `convertPorts()`, `getResourceReqs()` are used by the generators only and return kube objects so they will remain in generator utils. `addSyncRootFolder()`, `addSyncFolder()` are used by the generators only and will remain in generator utils.

I hope I was able to explain what my intentions are.

Reply:
Ok got it, thanks :+1: